Add json support, fix #10, ref XQF/xqf#12

illwieckz opened the pull request:
Hi @stevenh, this PR imports patches from the @eXprojects's qstat-json fork.

This fork was based on a source tarball, so was amnesic to all the revision history, also, this fork included mistakenly some temporary build files.

I've rewritten by hand the two useful patches and I committed them with the original `author` (`Steve Teuber <steve<ad>exprojects.org>`) and the original `author date` for each commit.

After that I've modified the source to add json support to games that were added since the release  tarball Steve used (like `starmade`, `dirtybomb`, `farmsim` and `ksp`) and fixed the `ventrilo` omission he did.

After that I renamed `name_xfroms` to `xform_names` and used xform_printf() instead of fprintf() in json code, like in e777674 (_Refactored output splitting out xform processing into its own files improving various portions of it._)

So, some intermediate commits cannot build, but the whole PR builds. I think we can't do better than that without rewriting Steve's commits, which raises legal issues: it must be clear to know who have done what and when.

@skybon, you will probably be interested by this PR.

skybon commented:
@illwieckz Nice work! Could you please rewrite the first two commits using @eXprojects current GitHub email? Thanks.

illwieckz commented:
I used the mail @eXprojects used himself, see [eXprojects/qstat-json/commit/1c56085.patch](https://github.com/eXprojects/qstat-json/commit/1c560859b52472d6b5638298f3b6b28fa8c66f90.patch) for example. Legally I think it's the better thing to do.

skybon commented:
@illwieckz GitHub does profile linking using email. Since first and foremost requirement is to make proper attribution I believe it is better to use @eXprojects' current email.

illwieckz commented:
I think the only user email Github reveals is the commit one.

skybon commented:
@illwieckz And that's the email those two commits should have too :-) Check out latest commits [here](https://github.com/AtomLinter/linter-puppet-lint)

illwieckz commented:
Well, the better is to ask @eXprojects himself if he wants I rewrite the commits whith his `Steve Teuber <steve.teuber@idealo.de>` author info.

illwieckz commented:
I've emailed Steve to ask him which address he wants (he can also just add the former address to his GitHub profile to get profile linking on that commit).

stevenh commented on qstat.c:
```diff
@@ -2523,54 +2530,968 @@ void xml_display_savage_player_info(struct qserver *server)
+	xform_printf(OF, "\t\t</players>\n");
+}
+
+/**
```
Can we split these out to a separate file please e.g. json.c, as qstat.c is already massive.

illwieckz replied:
**massively** done

stevenh commented on qstat.c:
```diff
@@ -2523,54 +2530,968 @@ void xml_display_savage_player_info(struct qserver *server)
+	server_type *type;
+	server_type **sorted_types;
+
+	sorted_types = (server_type **)malloc(sizeof(server_type*) * n_server_types);
```
Error checking for malloc failure

illwieckz replied:
will be rewritten with a function that knows how to list protocols from `qstat.cfg`.

stevenh commented on qstat.c:
```diff
@@ -2523,54 +2530,968 @@ void xml_display_savage_player_info(struct qserver *server)
+
+	sorted_types = (server_type **)malloc(sizeof(server_type*) * n_server_types);
+	type = &types[0];
+	for (i = 0; type->id != Q_UNKNOWN_TYPE; type++, i++)
```
iterator limit doesn't match malloc

illwieckz replied:
in fact this code does not list protocols from `qstat.cfg`, so it will be rewritten.

illwieckz replied:
erratum, this code must be able to list protocols from `qstat.cfg`, I'm bissecting to find regression.

illwieckz replied:
well no it works, it's just I'm tired.

illwieckz replied:
OK, @stevenh, this code was introduced by Steve Jankowski in the initial revision (2002-11-14), this code must be refactored, and refactoring will add XML support for this, but this will be a future pull request. This code is used since 13 years, so please do not prevent the PR merge about that.

illwieckz replied:
this is not in the scope of this PR, see #21 for a dedicated issue.

stevenh replied:
I can't accept a PR which introduces new issues, this needs to be fixed first

illwieckz replied:
This is not a new issue: https://github.com/multiplay/qstat/blob/fb2e96716e50438660085dfcc94eb9ee0f7520cd/qstat.c#L2655

This is legacy code. This PR can't rewrite the whole qstat code each time it reuses something.

A future PR can fix that. This code is legacy code which is running each time someone does `qstat --help` since 13 years. This fix is absolutely not in the scope of this PR.

stevenh replied:
You're adding this code in a new method, so my original comment still stands I can't accept a PR which clearly introduces new issues.

illwieckz replied:
Since master is now fixed, I have to rebase on master and import that fix.

stevenh commented on qstat.c:
```diff
@@ -2523,54 +2530,968 @@ void xml_display_savage_player_info(struct qserver *server)
+	{
+		if (!up_servers_only)
+		{
+			xform_printf(OF, "%s\n", (json_printed) ? ",\n\t{" : "\t{");
```
Use of %s for a const switch seems a bit wasteful

illwieckz replied:
We can do that?

```diff
<			xform_printf(OF, "%s\n", (json_printed) ? ",\n\t{" : "\t{"); 
>			xform_printf(OF, (json_printed) ? ",\n\t{\n" : "\t{\n"); 
```

stevenh replied:
Yup

illwieckz replied:
fixed

stevenh commented on qstat.c:
```diff
@@ -202,6 +202,9 @@ int no_port_offset = 0;
 int xml_display = 0;
 int xml_encoding = ENCODING_LATIN_1;
+int json_display = 0;
+int json_encoding = ENCODING_LATIN_1;
```
Does it not make more sense to have the default UTF8 for json?

illwieckz replied:
I don't know. @eXprojects , what do you think about it?

I'm not against that at all. In fact, a full UTF-8 output _whatever_ the format (including `raw` format) would be the better, there is already some issues I temporarily workaround in XQF (cf. [XQF/xqf:src/qstat.c:890](https://github.com/XQF/xqf/blob/6672d801da427dbc4cdb4769d636a8dff4b7398f/src/stat.c#L890) and [XQF/xqf:src/qstat.c:737](https://github.com/XQF/xqf/blob/6672d801da427dbc4cdb4769d636a8dff4b7398f/src/stat.c#L737)) because the `qstat` says `-utf8` is for `-xml` and XQF is using `-raw`.

Also, perhaps the issue #11 is related to the character encoding hell too.

illwieckz replied:
[RFC 4627](http://www.ietf.org/rfc/rfc4627.txt) says:

> JSON text SHALL be encoded in Unicode.  The default encoding is UTF-8.

UTF-8 is the only unicode encoding supported by `qstat`, so JSON display is restricted to UTF-8 encoding only.

fixed.

stevenh replied:
Where's the fix?

illwieckz replied:
there is another PR for fixes, after the split: #26

do you understand you're asking for maintaining one branch per commit, which is insane to track change and to maintain? If I move this fix to this PR, I need to rebase 5 branch (only 5 because I created less branch than commits), and because of the split you asked, the rebase will not be automatic?

illwieckz replied:
* There is a commit pending for this fix https://github.com/illwieckz/qstat/commit/73910f089d13c74fe898900242f89aec49a99c4e.
* You just have to pull the `json-improvement` branch: https://github.com/illwieckz/qstat/compare/split...illwieckz:json-improvement.
* There is a request for pull this branch: #26.

stevenh commented on qstat.c:
```diff
@@ -2523,54 +2530,968 @@ void xml_display_savage_player_info(struct qserver *server)
+		xform_printf(OF, "\t\t\t{\n");
+		xform_printf(OF, "\t\t\t\t\"name\": \"%s\",\n", json_escape(xform_name(player->name, server)));
+		xform_printf(OF, "\t\t\t\t\"score\": %d,\n", player->frags);
+		if ( - 999 != player->deaths)
```
style should be -999 != player->deaths

illwieckz replied:
You introduced the space between `-` and `999` in `xml_display_unreal_player_info` in 4e34cbf (2007-10-26), the `json` version is just a carbon copy. :wink:

illwieckz replied:
anyway it's fixed in f45100a

stevenh replied:
Said commit needs to added an squashed to this PR then.

stevenh commented on qstat.c:
```diff
@@ -2523,54 +2530,968 @@ void xml_display_savage_player_info(struct qserver *server)
+			xform_printf(OF, "\t\t\t\t\"team\": %d,\n", player->team);
+		}
+
+		// Some games dont provide
```
Comment is pretty none de-script, I get the meaning but we should make it better or remove it as its really quite obvious from the code thats whats happening.

illwieckz replied:
fixed

stevenh replied:
Where?

stevenh commented on qstat.c:
```diff
@@ -2523,54 +2530,968 @@ void xml_display_savage_player_info(struct qserver *server)
+
+		// Some games dont provide
+		// so only display if they do
+		if (player->skin)
```
pointer tests should with with null so its clear

illwieckz replied:
fixed

stevenh commented on qstat.c:
```diff
@@ -2523,54 +2530,968 @@ void xml_display_savage_player_info(struct qserver *server)
+	player = server->players;
+	for (; player != NULL; player = player->next)
+	{
+		if (printed) xform_printf(OF, ",\n");
```
style, avoid using single line if's like this.

illwieckz replied:
fixed

stevenh replied:
I'm not seeing it fixed here.

stevenh commented on qstat.c:
```diff
@@ -2523,54 +2530,968 @@ void xml_display_savage_player_info(struct qserver *server)
+	{
+		if (printed) xform_printf(OF, ",\n");
+		xform_printf(OF, "\t\t\t{\n");
+		xform_printf(OF, "\t\t\t\t\"name\": \"%s\",\n", json_escape(xform_name(player->name, server)));
```
Helper functions for outputting single json value could make this quite a bit cleaner something along the lines of
```c
json_output_string(int lvl, char *name, char *val)
{
....
}
```

illwieckz replied:
could be done in another pull request in the future, same for xml

stevenh replied:
Agreed

stevenh commented on qstat.c:
```diff
@@ -2523,54 +2530,968 @@ void xml_display_savage_player_info(struct qserver *server)
+	xform_printf(OF, "\n\t\t]");
 }
 
-void xml_display_farcry_player_info(struct qserver *server)
```
Looks like there may be a miss-merge here

illwieckz replied:
oh yes.

illwieckz replied:
fixed

stevenh replied:
Still broken here.

illwieckz replied:
see my other comment, the git diff tool is weird

stevenh commented on qstat.c:
```diff
@@ -3356,6 +4278,14 @@ int main(int argc, char *argv[])
 		{
 			usage(NULL, argv, NULL);
 		}
+		else if (strcmp(argv[arg], "--json-protocols") == 0)
```
Please keep command line options in the standard style

illwieckz replied:
fixed

stevenh replied:
Still looks like --option vs -option here

illwieckz replied:
This thing is done in another PR that revamps the help print. Is it relevant to modify this line since this line will completely disappear?

stevenh commented on qstat.c:
```diff
@@ -3488,6 +4418,14 @@ int main(int argc, char *argv[])
 				usage("cannot specify both -raw and -xml\n", argv, NULL);
 			}
 		}
+		else if (strcmp(argv[arg], "-json") == 0)
```
Should also conflict with xml and vice-versa

illwieckz replied:
fixed.

Just for information, the code was conflicting `-raw -xml`, but not `-xml -raw`. This is fixed too.

stevenh replied:
Not seeing a fix

illwieckz replied:
Same, this issue is not the only like that, another PR fixes that and other ones. It's just too much to do all in the current PR.

stevenh commented on qstat.c:
```diff
@@ -12385,6 +13331,225 @@ char *xml_escape(char *string)
+		c = *string;
+		switch (c)
+		{
+			/*case '&':
```
Remove commented out code

illwieckz replied:
fixed

stevenh commented on qstat.c:
```diff
@@ -12385,6 +13331,225 @@ char *xml_escape(char *string)
 
+char *json_escape(char *string)
+{
+	static unsigned char _buf[4][MAXSTRLEN + 8];
```
The use of this sort of round robin buffering is asking for trouble.

illwieckz replied:
@stevenh , this code was in the initial revision by Steve Jankowski (2002-11-14) with the form `static unsigned char _buf[4][MAXSTRLEN];` and modified by Steve Jankowski (2003-02-23) with the form `static unsigned char _buf[4][MAXSTRLEN+8];`, then modified by you (2007-10-26) with the form `static unsigned char _buf[4][MAXSTRLEN + 8];`.

Please be indulgent, I hop this line will not prevent the merge, this line is used as-is without known bug since 2003.

This improvement could be made later.

An xml/json refactoring could be made later, but it's too heavy for this pull request. This PR was intended to add JSON support, not to rewrite the XML support. So, a future PR can handle that, and this line does not prevent this PR to be merged.

illwieckz replied:
this is not in the scope of this PR, see #22 for a dedicated issue.

stevenh replied:
Your adding this code in the PR which could easily cause problems, as previously I can't accept a PR which has obvious potential issues.

Comparing to previous commits which have the same issues is irrelevant I'm afraid.

stevenh commented on qstat.h:
```diff
@@ -386,6 +386,39 @@ void xml_display_ventrilo_player_info( struct qserver *server);
 void xml_display_starmade_player_info( struct qserver *server);
 char *xml_escape( char*);
 
+void json_display_server( struct qserver *server);
```
Move to separate header to avoid more bloating of this already huge header.

illwieckz replied:
fixed

stevenh commented on qstat.h:
```diff
@@ -386,6 +386,39 @@ void xml_display_ventrilo_player_info( struct qserver *server);
+void json_display_server( struct qserver *server);
+void json_header();
+void json_footer();
+void json_display_server_rules( struct qserver *server);
```
Style remove leading space after ('s

illwieckz replied:
fixed

stevenh commented on qstat.h:
```diff
@@ -3200,6 +3369,8 @@ server_type builtin_types[] = {
     raw_display_server_rules,			/* display_raw_rule_func */
     xml_display_ventrilo_player_info,	/* display_xml_player_func */
     xml_display_server_rules,			/* display_xml_rule_func */
+    json_display_ventrilo_player_info,	/* display_json_player_func */
```
indentation issue?

illwieckz replied:
This whole file is misindented, is there somewhere a documentation file that describe the indentation guideline?

The former code was not following the current indentation (use tab instead of space and vice-versa) and was badly aligned, what your are seeing is something I have _already reindented, by hand, line per line_.

The current code use this rule:

> * use 4-spaces for code indentation (before code)
> * use tab for comment alignment (after code), sometime the tab must be 8-space long, sometime 4, it depend where you are in the same file.

Github use 8-char-long space and qstat mixes tab and spaces and not use a 8-char-long indentation, so Github can't render the qstat code alignment.

So, the current code use space before code and tab after code, and some alignment are ok with a 4-char-long tab and some alignment are ok with a 8-char-long tab. If the code must be reindented I think the 4-char-long tab is better since it's the spacing used for blocks.

For information, in XQF I use tab for indentation and space for ascii art, a comment is considered as ascii art (so, even if the tab size is variable, the comment is always aligned).

illwieckz replied:
fixed, with 4-char tab, since the code used 4-char tabs before

stevenh commented on qstat.h:
```diff
@@ -3234,6 +3405,8 @@ server_type builtin_types[] = {
     raw_display_server_rules,			/* display_raw_rule_func */
     NULL,								/* display_xml_player_func */
     xml_display_server_rules,			/* display_xml_rule_func */
+    NULL,                               /* display_json_player_func */
```
indentation issue?

illwieckz replied:
fixed

stevenh commented on qstat.h:
```diff
@@ -3505,6 +3692,8 @@ server_type builtin_types[] = {
     NULL,			/* display_raw_rule_func */
     NULL,			/* display_xml_player_func */
     NULL,			/* display_xml_rule_func */
+    NULL,           /* display_json_player_func */
```
indentation issue?

illwieckz replied:
Yes, it's something I forget to rewrite. There is space before and after `NULL` but the other lines use space before and tabs after.

Qstat really needs to unify the style.

stevenh replied:
Yes it does ;-)

illwieckz replied:
fixed

illwieckz commented:
so, the code now splits `qstat.c` to `output-standard.c`, `output-xml.c` and `output-json.c` (all with a `.h` file).

`template.c` was renamed to `output-template.c` and the missing `output-template.h` was created, and `output-template.h` now rely on `qstat.h` (before that, some parts were copypasted in `template.c`).

illwieckz commented:
So there is more code rewritten and cleaned than added, it's more a cleaning effort than a code addition…

illwieckz commented:
@stevenh, can you help me on these parts:
https://github.com/multiplay/qstat/pull/13#discussion-diff-33666273
https://github.com/multiplay/qstat/pull/13#discussion-diff-33666396
https://github.com/multiplay/qstat/pull/13#discussion-diff-33667166

:smiley: ?

illwieckz commented:
Ok, I've added the `-nx`/`-nnx` options to qstat usage print, and removed one of the two `-nh` indications. I also unified the uppercase/lowercase usage description, then I wrote a `printf_opt` function to align options properly, then I rearranged the usage display. Now it looks like that:

```
$ ./qstat
Usage: ./qstat [options] [-default server-type] [-cfg file] [-f file] [host[:port]] ...
Where host is an IP address or host name

Configuration options:
    -cfg                 Read the extended types from given file not the default one
    -nocfg               Ignore qstat configuration loaded from any default location. Must be the first option on the command-line.

Game query options:
    -a2s                 Query Half-Life 2 new server
    -ams                 Query America's Army v2.x server
    -bfbc2               Query Battlefield Bad Company 2 server
    -bfs                 Query BFRIS server
    -crysis              Query Crysis server
    -cubes               Query Sauerbraten server
    -d3g                 Query Descent3 Gamespy Protocol server
    -d3m                 Query Descent3 Master (PXO) server
    -d3p                 Query Descent3 PXO protocol server
    -d3s                 Query Descent3 server
    -dirtybomb           Query DirtyBomb server
    -dm3m                Query Doom 3 Master server
    -dm3s                Query Doom 3 server
    -efm                 Query Star Trek: Elite Force server
    -efs                 Query Star Trek: Elite Force server
    -etqws               Query QuakeWars server
    -eye                 Query All Seeing Eye Protocol server
    -farmsim             Query FarmingSimulator server
    -fcs                 Query FarCry server
    -fls                 Query Frontlines-Fuel of War server
    -gps                 Query Gamespy Protocol server
    -grs                 Query Ghost Recon server
    -gs2                 Query Gamespy V2 Protocol server
    -gs3                 Query Gamespy V3 Protocol server
    -gs4                 Query Gamespy V4 Protocol server
    -gsm                 Query Gamespy Master server
    -h2s                 Query Hexen II server
    -hazes               Query Haze Protocol server
    -hl2s                Query Half-Life 2 server
    -hlm                 Query Half-Life Master server
    -hls                 Query Half-Life server
    -hrs                 Query Heretic II server
    -hwm                 Query HexenWorld Master server
    -hws                 Query HexenWorld server
    -jk3m                Query Jedi Knight: Jedi Academy server
    -jk3s                Query Jedi Knight: Jedi Academy server
    -kps                 Query Kingpin server
    -ksp                 Query Kerbal Space Program server
    -mumble              Query Mumble server
    -ottdm               Query openTTD Master server
    -ottds               Query OpenTTD server
    -preys               Query PREY server
    -prs                 Query Pariah server
    -q2m                 Query Quake II Master server
    -q2s                 Query Quake II server
    -q3m                 Query Quake III Master server
    -q3s                 Query Quake III: Arena server
    -q4m                 Query Quake 4 Master server
    -q4s                 Query Quake 4 server
    -qs                  Query Quake server
    -qwm                 Query QuakeWorld Master server
    -qws                 Query QuakeWorld server
    -rss                 Query Ravenshield server
    -rwm                 Query Return to Castle Wolfenstein Master server
    -rws                 Query Return to Castle Wolfenstein server
    -sas                 Query Savage server
    -sfs                 Query Soldier of Fortune server
    -sgs                 Query Shogo: Mobile Armor Division server
    -sns                 Query Sin server
    -starmade            Query StarMade server
    -stm                 Query Steam Master server
    -t2m                 Query Tribes 2 Master server
    -t2s                 Query Tribes 2 server
    -tbm                 Query Tribes Master server
    -tbs                 Query Tribes server
    -tee                 Query Teeworlds server
    -terraria            Query Terraria server
    -tm                  Query TrackMania server
    -ts2                 Query Teamspeak 2 server
    -ts3                 Query Teamspeak 3 server
    -uns                 Query Unreal server
    -ut2004m             Query UT2004 Master server
    -ut2s                Query Unreal Tournament 2003 server
    -vent                Query Ventrilo server
    -wics                Query World in Conflict server
    -wolfs               Query Wolfenstein server
    -default             Set default server type

Server types: a2s ams bfbc2 bfs crysis cube2 d3g d3m d3p d3s dirtybomb dm3m dm3s efm efs etqws eye farmsim fcs fls gps grs gs2 gs3 gs4 gsm h2s hazes hl2s hlm hls hrs hwm hws jk3m jk3s kps ksp mumble ottdm ottds preys prs q2m q2s q3m q3s q4m q4s qs qwm qws rss rwm rws sas sfs sgs sns starmade stm t2m t2s tbm tbs tee terraria tm ts2 ts3 uns ut2004m ut2s ventrilo wics wolfs

Content options:
    -sort                Sort servers and/or players
    -u                   Only display servers that are up
    -nf                  Do not display full servers
    -ne                  Do not display empty servers
    -R                   Fetch and display server rules
    -P                   Fetch and display player info

Format options:
    -cn                  Display color names instead of numbers
    -ncn                 Display color numbers instead of names
    -hc                  Display colors in #rrggbb format
    -htmlmode            Convert <, >, and & to the equivalent HTML entities
    -htmlnames           Colorize Quake 3 and Tribes 2 player names using html font tags
    -nohtmlnames         Do not colorize Quake 3 and Tribes 2 player names even if $HTML is used in an output template.
    -carets              Display carets in Quake 3 player names
    -hpn                 Display player names in hex
    -hsn                 Display server names in hex
    -tc                  Display time in clock format (DhDDmDDs)
    -tsw                 Display time in stop-watch format (DD:DD:DD)
    -ts                  Display time in seconds
    -pa                  Display player address

Display options:
    -raw <delim>         Output in raw format using <delim> as delimiter
    -raw-arg             When used with -raw, always display the server address as it appeared in a file or on the command-line.
    -old                 Old style display
    -nh                  Do not display header
    -nx                  Enable name transformations, -nx is set by default
    -nnx                 Disable name transformations, -utf8 option implies -nnx
    -xml                 Output status data as an XML document
    -bom                 Output Byte-Order-Mark for XML output.
    -utf8                Use the UTF-8 character encoding for XML output
    -json                Output status data as an UTF-8 JSON document
    -Th,-Ts,-Tpt         Output templates: header, server and player
    -Tr,-Tt              Output templates: rule, and trailer
    -showgameport        Always display the game port in QStat output.
    -stripunprintable    Disable stripping of unprintable characters.
    -errors              Display errors
    -progress            Display progress meter on stderr (text only)

Output options:
    -of                  Output file
    -af                  Like -of, but append to the file

Query options:
    -f                   Read hosts from file
    -mdelim <delim>      For rules with multi values use <delim> as delimiter
    -retry               Number of retries, default is 3
    -interval            Interval between retries, default is 0.50 seconds
    -mi                  Interval between master server retries, default is 2.00 seconds
    -timeout             Total time in seconds before giving up
    -maxsim              Set maximum simultaneous queries
    -sendinterval        Set time in ms between sending packets, default 5
    -allowserverdups     Allow adding multiple servers with same ip:port (needed for ts2)
    -srcport <range>     Send packets from these network ports
    -srcip <IP>          Send packets using this IP address
    -H                   Resolve host names
    -Hcache              Host name cache file

Advanced options:
    -d                   Enable debug options. Specify multiple times to increase debug level.
    -dump                Write received raw packets to dumpNNN files which must not exist before
    -pkt <file>          Use file as server reply instead of quering the server. Works only with TF_SINGLE_QUERY servers
    -syncconnect         Process connect initialisation synchronously.
    -noportoffset        Dont use builtin status port offsets (assume query port was specified).

Sort keys:
  servers: p=by-ping, g=by-game, i=by-IP-address, h=by-hostname, n=by-#-players, l=by-list-order
  players: P=by-ping, F=by-frags, T=by-team, N=by-name

qstat version 2.15
```

illwieckz commented:
OK, the only thing to fix is the protocol list, theres is already one in the `usage()` function.

illwieckz commented:
@eXprojects, to query `qstat` version with json display, you must do that now:

```sh
qstat -json -v
```

to query protocols, now you must do that now:

```sh
qstat -json -protocols
```

illwieckz commented:
@stevenh, OK this, PR is massive and it can be merged now, I hope you will be able to merge it soon, and I hope you will like all the improvements! :smiley:

illwieckz commented:
The teeworld PR (#3) will be rebased after this PR merge, so, there is  no reason to wait more.

stevenh commented:
Sorry to be a pain but fixing style of "existing" code and splitting out other sections should be done in a separate PR so its easy to catch issues. As this stands currently its very hard to see what's new and what's just moved code.

illwieckz commented:
> Add to that unrelated other fixes e.g. illwieckz@b1a54b6

When you are doing things like this PR does, you must be sure there is no new warning introduced, and the better way is to have a warningless compile, so every warning is a new.

In fact, my first attempt of this commit https://github.com/illwieckz/qstat/commit/c53e2948477d3f0d1c0b2a54c1ad43f9fa69e69c introduced some mistakes that appeared as warning, I fixed the warnings to be able to fix my commit.

You can't fix code if you keep mistakes.

illwieckz commented:
As you can see in the history, I wrote the warning-cleaning commit 5 commit after the NULL pointer commit, it means because of this warning noise I've wrote 4 commits before seeing I've introduced a regression.

illwieckz commented:
> Sorry to be a pain but fixing style of "existing" code and splitting out other sections should be done in a separate PR so its easy to catch issues.

This is why I wrote different commits, after the merge, the PR will be nothing, catching issue is about commit.

Also, because of this work is huge, it would have been very difficult without cleaning some code to help to read the code.

The code can't be maintained if it's not readable.

stevenh commented:
As a general rule each PR should be an individual commit, re-factoring / formatting code unrelated to the PR is a big no no.

Fixes made due to PR feedback should be squashed to avoid history bloat.

illwieckz commented:
> As this stands currently its very hard to see what's new and what's just moved code.

There is different commits for what is new, what is moved code, and what is spacing. Also, there is many standard git tools that allows every one to view diffs without spaces changes.

For example with commit https://github.com/illwieckz/qstat/commit/1d9d6c552cadb595eca46fdc4641c00ca9c3f7be that reindent qstat.cfg with about 3000 lines modified, the only not-spacing change is:

```diff
$ git show --ignore-all-space 1d9d6c552cadb595eca46fdc4641c00ca9c3f7be
commit 1d9d6c552cadb595eca46fdc4641c00ca9c3f7be
Author: Thomas Debesse <dev@illwieckz.net>
Date:   Sun Jul 5 08:30:31 2015 +0200

    reindent qstat.h, ref #14

diff --git a/qstat.h b/qstat.h
index 74da23d..4cf303b 100644
--- a/qstat.h
+++ b/qstat.h
@@ -877,7 +877,8 @@ char terraria_serverstatus[] = "GET /v2/server/status HTTP/1.1\x0d\x0a\x0d\x0a";
 /* SERVER BUILTIN TYPES */
 
 
-server_type builtin_types[] = {
+server_type builtin_types[] =
+{
        {
                /* QUAKE */
                Q_SERVER,                                               
```

And in fact it's spacing change too, this is just the diff does not count new lines as space.

illwieckz commented:
Well, this PR was just a
> As a general rule each PR should be an individual commit

It's not true, a PR must not have only one commit, it must be coherent, if there is too much changes, they must be dispatched in different atomic commits to avoid jumbo commits.

This kind of commit is not acceptable: https://github.com/multiplay/qstat/commit/e777674c9d287af6ad1cd470237567e9b50f3652 it's from you, it's called _"Refactored output splitting out xform processing into its own files improving various portions of it"_ but it's not only refactor output, it adds support for 3 games, changes packet handling and bump version. All these changes should have been splitted in different commits…

This PR is like the tee server PR, I've just wanted to fix some things that were wrong, but in the end I have rewritten all the whole file.

About this PR, All I wanted was to merge code from another guy, code I did not write myself.

You asked to rewrite legacy qstat code I never wrote and which is in VCS since more than 10 years or since the beginning.

You asked for reformatting code I never wrote, (like qstat.h code).

You asked to split the json display code, but the better way to achieve this is to display the whole display code, since it's the best way to do it properly with kinds `.h` files to not declare the same things multiple time.

> formatting code unrelated to the PR is a big no no.

You asked to split code and to reformat code. The split can't be made cleanly without reformatting. The split can't be made cleanly without changing files unrelated to the PR. The split can't be made cleanly without reformatting - bis.

illwieckz commented:
> Fixes made due to PR feedback should be squashed to avoid history bloat.

Well, each commit have been squashed 20 time.

The only commits I never squashed are those ones that are not mine, because I can't legally do that.

All the commits are atomic (this is why this PR have multiple commits) except those ones that are not mine, because I can't legally change them.

All commits build, except those ones that are not mine, because I can't legally change them.

illwieckz commented:
Well, each time I make a PR you ask for things that need to reformat and rewrite all, including legacy qstat code, and after that you said « it's change too much things, it's a big no-no ». What is the problem?

If you need 1 commit per PR, just merge the PR as-is when he comes with only one commit, do not ask to unrelated change that must not be squashed, and submit an issue to ask another PR for the other commits.

But I think you do not want 19 PR for 19 commits.

illwieckz commented:
@eXprojects: do you allow me to squash your two commits named "Added json support" and "fixed exit code" ?

stevenh commented:
> As a general rule each PR should be an individual commit
>> It's not true, a PR must not have only one commit, it must be coherent, if there is too much changes, they must be dispatched in different atomic commits to avoid jumbo commits.

Each project / team has their own rules, this is one of ours, if a change is so big it would need multiple commits, its likely the change is too complex and should be split anyway. As I said its a general rule, not hard and fast.

The key is not to muddy the water with fixes required to allow a change to get committed. So if additional changes / corrections are made due to feedback then these should be squashed. I'm happy for the third party commits to remain as is.

> This kind of commit is not acceptable: e777674 it's from you,...

What's in the past is just that.

> You asked for reformatting code I never wrote, (like qstat.h code).

I asked that for the code you wanted to add, not reformat and split everything, sorry if that wasn't clear.

Its important moving forward that the changes are clear as this makes them much easier to review and test. In its current form splitting out all the raw methods etc makes it very hard to see what the actual functional changes are.

Hope that clarifies.

illwieckz commented:
I asked that for the code you wanted to add, not reformat and split everything, sorry if that wasn't clear.

The point is the code I wanted to add was just copy pasting from legacy code, so changing this code is changing the legacy qstat too.

The best way to fix code you pointed out would be to refactor it to be able to fix it only once.

Since the refactoring is a too heavy change for this PR, I choose to fix code you pointed out and to fix other occurrence of this code to keep the code base coherent and to help a future refactoring to be doable.

About the raw and xml split, splitting the JSON code without splitting other display code is just the best way to fall in multiple #include hell, and it's the best way to merge ugly hacks and unmaintainable and incoherent code. The split should have been asked as another PR from the beginning.

I've just did what you asked, but the cleaner way, and the cleaner way needs more rewrite than the fast way.

If you do not like the cleaner way and prefer the fast-and-dirty way, why you do not merge the PR as-is and as fast as it comes, since the cleaning must be done in another PR ?

It seems you just do not measure the exorbitancy work you ask, and you did not notice that fixes you ask is mainly about legacy code. So, it's normal if the fixes asked just rewrite all the legacy code.

illwieckz commented:
> In its current form splitting out all the raw methods etc makes it very hard to see what the actual functional changes are.

The splitting just split the files, it does not change the functional code (except some `#include` and `Makefile.am` lines of course).

This is why, for example, I've not refactored both `json_escape` and `xml_escape` yet since refactoring is changing functionnal code. So, when I fix `display_json.c` because you ask it, I fix the exact same code in `display_xml.c` too to keep the code coherent. If I don't do that, who knows what is the good code and what is the dirty code if we keep both the good and the dirty code?

Also, if you ask to fix a spacing in one line of a 4000 line file, I will not just fix this line to be incoherent with the 3999 other lines, fixing a spacing in one line of a 4000 line file can't be done cleanly without fixing the other 3999 of the same file.

So, in end, the only way to do it cleanly what you ask is to do it * 4000.

illwieckz commented:
Another example, fixing cleanly the `--json-protocols` is not to rename this option to `-jsonprotocols`, but to allow to do `-json -protocols` the same way we already do `-json -q3m`, so, the clean way to fix one line is to change other lines.

Almost all the comments you made do not take into account the existing code. So, you do not expect the change you ask.

illwieckz commented:
I'm splitting the PR, but I will not do one PR per commit and potentially obligate me to rebase 20 branches if a potential change needs that.

illwieckz commented:
So, you have now these PR:

* PR #13 _Add json support, fix #10, ref XQF/xqf#12_
* PR #23 _Split the display code._
* PR #24 _Begin to clean style, spacing an indentation, ref #14_
* PR #25 _Unset unused variables to avoid warnings, ref #19_
* PR #26 _Json improvement, ref #13, ref #10, ref XQF/xqf#12_
* PR #27 _Qstat help improvement_

I can't move #26 after #13 since it needs #23, #24, #25 to be achieved. I really needed all these splitting and cleaning effort to be able to write #26.

The json improvements you asked for needs these splitting and cleaning effort to be achieved first. They are just improvements to #13 which can be merged without #26 since it works without #26.

Rewriting one commit in on PR that another PR depends on will be very painfull (rebasing up to 5 PR).

illwieckz commented:
Just for information, Github will not displays a clean N PR diff until you don't merge N-1 PR, it's normal, because Github does not know if a PR depends on another, so each PR recapitulates the diff from `master`.

illwieckz commented:
Just one last thing about PR:

> As a general rule each PR should be an individual commit
> Each project / team has their own rules

There is no PR in git, after a merge, a PR is nothing about a noisy additional commit in the worst case (like GitHub does), PR is just a social mechanism to say « hello, I have changes for you ».

A _pull request_ is just the name GitHub gives to the process to request a pull from a foreign branch, some other online hosting platform call it a _merge request_ for example, and in the traditional form, it's just a _request for pull_, like there is _request for comments_.

A pull request is nothing than requesting a pull from a foreign branch, git does not know what is a PR, and there is absolutely no rule in Git that said there must be only one commit per branch, because there is already two obvious rule: _there is one commit per commit_, and _there is _branches for multiples commits for changes that does not fit in one commit_.

Git was not created to have only one branch per commit, which is insane to maintain, and a request for pull is about branch… If a PR were about commits and not branch, it were not called a _pull request_ but a _cherry-pick request_…

Foreign commits are not pulled, they are cherry-picked, pull is for branches, branches is for multiple commit about the same effort, and jumbo commits that made too much things must be splitted, but branch with multiple commits designed for a unique purpose must not be splitted.

illwieckz commented:
Just to help you:

* reviewing the PR #13 _Add json support, fix #10, ref XQF/xqf#12_:  
https://github.com/multiplay/qstat/compare/master...illwieckz:json
* reviewing the PR #23 _Split the display code, fix #28_:  
https://github.com/illwieckz/qstat/compare/json...illwieckz:split-display
* reviewing the PR #24 _Begin to clean style, spacing an indentation, ref #14_:  
https://github.com/illwieckz/qstat/compare/split-display...illwieckz:clean-spacing
* reviewing the PR #25 _Unset unused variables to avoid warnings, ref #19_:  
https://github.com/illwieckz/qstat/compare/clean-spacing...illwieckz:unset-unused
* reviewing the PR #26 _Json improvement, ref #13, ref #10, ref XQF/xqf#12_:  
https://github.com/illwieckz/qstat/compare/unset-unused...illwieckz:json-improvement
* reviewing the PR #27 _Qstat help improvement, fix #17_:  
https://github.com/illwieckz/qstat/compare/json-improvement...illwieckz:qstat-help

stevenh commented:
Making progress, but the issues highlighted by the review need to be fixed in this PR and not in separate PR's to ensure the code base is in good state no matter which revision is used.

Its great to use new PR's for other issues like general style cleanup and display split which effect other areas of the code but each new PR needs to pass review on its own prior to commit.

Currently your PR's are all dependent on each other instead of being based off separate branches, which make them impossible to review :(

By altering the structure of how they are committed it should be possible to all but eliminate these interdependencies.

From what I can determine we should see the following PR's:
1. Eliminate unused vars (commits from PR #25)
2. Add json support (combine PR #13, PR #26, PR #27)
3. Style cleanup
4. Split existing display code out to separate files.

Keep up the good work :)

illwieckz commented:
> 1. Eliminate unused vars (commits from PR #25)
> 2. Add json support (combine PR #13, PR #26, PR #27)
> 3. Style cleanup
> 4. Split existing display code out to separate files.

I can't do that, it means redoing all the work from scratch, with risk to introduce new issues.

The work was done in the order I use because it was the safest way to do this job.

I can only do that:

> 1. Add json support (+ fix utf8 etc. from PR #26), it means backports all the change by hand from #26 then rebase the next branch by hand (git will be able to rebase the others automatically but not this one).
> 2. Split the display code.
> 3. Begin to clean style - depend of the split
> 4. Unset unused variables - depend of the cleaning
> 5. Qstat help improvement + args must exclude themselves. This is not a JSON issue, it's just the current JSON effort spotted it, there was already a problem with xml and raw, it's better to keep this fix as another PR, it's not related to JSON processing, it's related to Qstat user interface.

2. 3. 4. and 5. are not JSON-related, it's just some tasks motivated by JSON, they can be done in the order we want, they change nothing for qstat and for you. But it change many thing to me if we do not do that in the order above, because it means to me redoing all from scratch.

illwieckz commented on qstat.c:
```diff
@@ -2478,99 +2485,1013 @@ void xml_display_ventrilo_player_info(struct qserver *server)
+	xform_printf(OF, "\t\t</players>\n");
+}
+
+void xml_display_farcry_player_info(struct qserver *server)
```
github is fooled, it's already fixed

illwieckz commented:
Ok, this was was painful but it's done, it needed to manually backports changes from #26 to #13, #23 and #24. No branch was able to rebase automatically, but I was able to use merge/cherry-pick to recreate each branch without editing files (except for the backports from #26).

Merging the unused var PR #25 before the json pr #13 is undoable, it means backporting by hand all the changes to all other PR since merge will be broken, editing plenty of files by hand each time to just redoing the job in each history. It's the better way to introduce new issues and it is very unsafe and painful. The unused var issue is in the code since years, it's not a problem at all if it's fixed after the json support addition.

Fixing unused var before or after introducing json support is the same to you (json does not depend on it): it was not fixed, it will be fixed, it's ok, it's just another PR unrelated that can be merged right after others PR unrelated.

But fixing unused var before or after introducing json support is not the same to me, it means editing manually each branch for each other PR. It's unsafe for no gain.

This is now the PR list:

#13 _Add json support_
#23 _Split the display code_
#24 _Begin to clean style, spacing, indentation_ (cleaning display code + qstat.cfg)
#25 _Unset unused variables_
#27 _Qstat help improvement_, also fix the user interface (some options must exclude themselves).

PR #26 is now unneeded since all commits were backported to #13 or merged to #27.

illwieckz commented:
I've done some extra work with uncrustify to clean the source and I get very good results, so the PR #24 will just be pure noise in the history since I will submit a better one soon.

So, I closed the PR #24 and rebuilded the PR #25 and #27 to be able to merge them without it.

This is now the PR list:

#13 _Add json support_
#23 _Split the display code_
#25 _Unset unused variables_
#27 _Qstat help improvement_, also fix the user interface (some options must exclude themselves) and some query style.

skybon commented:
@stevenh Which issues preventing this PR from being merged remain? I'd be happy to help @illwieckz  with fixing those and making QStat JSON support happen asap :smile:

z commented:
What's the status on this?

illwieckz commented:
Hello @stevenh, I reedited all my PR to include the iterator limit to match malloc in protocol listing, I reminds you the merge order:

* #13 Add json support (the current one)
* #23 Split the display code
* #25 Unset unused variables
* #27 Qstat help improvement, also fixes the user interface (some options must exclude themselves) and some query style

illwieckz commented:
Hi @stevenh , is this PR ok for you now? I hope it will be merged soon.

stevenh commented:
Still looks like a number issues already haven't been addressed in this PR.

A secondary PR to fix them is an option as this means we'll have bad code in the repo at the point this is merged, so I'm afraid I'm going to have to insist all issues are addressed before we can get this merged.

illwieckz commented:
Hi,

> Still looks like a number issues already raised, haven't been addressed in this PR.

Please list them, _"looks like"_ is not an argument.

illwieckz commented:
Hi, I fixed in this PR the temporary double hyphen `--json` options wrote by eXprojetcs that are deleted by myself in PR #27 dedicated to option/arg handling fixes which is blocked by the current PR.

illwieckz commented:
Hi @stevenh , is there still something preventing the merge ?

stevenh commented:
merged via #53
